File: packages/guard/src/core/guard.ts
import { InMemoryStore } from "../anomaly/store.js";
import { getDefaultAnomalyConfig, scoreAnomaly } from "../anomaly/scorer.js";
import type {
  ChainOfCommandAuthorizationInput,
  ChainOfCommandDecision,
  ChainOfCommandReviewResponse,
  ChainOfCommandReviewTrailEntry,
  Decision,
  GuardConfig,
  GuardDecisionResult,
  GuardEngine,
  ToolCallContext,
  WrapToolOptions
} from "../types/index.js";
import {
  chainOfCommandEnabled,
  chainOfCommandMaxLevels,
  consumePermit,
  createPermitRecord,
  isUnsupportedByPolicy,
  unsupportedCallFingerprint,
  writePermit
} from "./chain-of-command.js";
import { decisionFromFindings, mergeDecisionWithAnomaly } from "./decision.js";
import { evaluatePolicy } from "./evaluate.js";
import {
  GuardApprovalDeniedError,
  GuardApprovalRequiredError,
  GuardBlockedError
} from "./errors.js";
import { normalizeToolCall } from "./normalize.js";

function fallbackAction(policyDefault: Decision | undefined): Decision {
  return policyDefault ?? "allow";
}

function decisionFromReviewResponse(response: ChainOfCommandReviewResponse): ChainOfCommandDecision | undefined {
  const decision = response.decision;
  if (decision === "yes" || decision === "no" || decision === "escalate") return decision;
  return undefined;
}

function initialSupervisorId(context: ToolCallContext): string {
  const metadata = context.metadata;
  const value = metadata?.chainOfCommandSupervisorId;
  if (typeof value !== "string" || value.trim().length === 0) {
    return "supervisor-level-1";
  }
  return value.trim();
}

function deniedChainResult(
  result: GuardDecisionResult,
  reviewTrail: ChainOfCommandReviewTrailEntry[]
): GuardDecisionResult {
  return {
    ...result,
    chainOfCommand: {
      ...result.chainOfCommand,
      status: "denied",
      reviewTrail: reviewTrail.map((entry) => ({ ...entry }))
    }
  };
}

function assertTerminalEvidence(response: ChainOfCommandReviewResponse): {
  reviewerId: string;
  reason: string;
  supervisorSignature?: string;
} | null {
  const reviewerId = response.reviewerId?.trim();
  const reason = response.reason?.trim();

  if (!reviewerId || !reason) {
    return null;
  }

  return {
    reviewerId,
    reason,
    supervisorSignature: response.supervisorSignature
  };
}

export function createGuard(config: GuardConfig): GuardEngine {
  const policy = config.policy;
  const store = config.store ?? new InMemoryStore();
  const anomalyConfig = {
    ...getDefaultAnomalyConfig(),
    ...(config.anomaly ?? {})
  };
  const chainEnabled = chainOfCommandEnabled(config.chainOfCommand?.enabled);
  const maxEscalationLevels = chainOfCommandMaxLevels(config.chainOfCommand?.maxEscalationLevels);

  const engine: GuardEngine = {
    async beforeToolCall(context: ToolCallContext): Promise<GuardDecisionResult> {
      const started = performance.now();
      const event = normalizeToolCall(context);
      const findings = evaluatePolicy(event, policy);
      const defaultAction = fallbackAction(policy.defaults.action);
      const policyDecision = decisionFromFindings(findings, defaultAction);
      const unsupportedByPolicy = isUnsupportedByPolicy({
        fallbackAction: defaultAction,
        findingsCount: findings.length,
        policyDecision
      });

      const fingerprint = unsupportedByPolicy ? unsupportedCallFingerprint(event) : undefined;
      let effectivePolicyDecision = policyDecision;
      let chainOfCommand: GuardDecisionResult["chainOfCommand"] = {
        status: "not_applicable",
        reviewTrail: []
      };

      if (chainEnabled && unsupportedByPolicy && fingerprint) {
        const permit = await consumePermit(
          store,
          { actorId: event.actorId, sessionId: event.sessionId },
          fingerprint
        );

        if (permit) {
          effectivePolicyDecision = "allow";
          chainOfCommand = {
            status: "approved_once",
            fingerprint,
            permitId: permit.permitId,
            reviewTrail: permit.reviewTrail.map((entry) => ({ ...entry }))
          };
        } else {
          chainOfCommand = {
            status: "eligible",
            fingerprint,
            reviewTrail: []
          };
        }
      }

      const anomaly = await scoreAnomaly(event, policy, store, anomalyConfig);

      const merged = mergeDecisionWithAnomaly(effectivePolicyDecision, anomaly.proposedAction);
      let decision = merged.decision;
      let escalatedByAnomaly = merged.escalatedByAnomaly;

      if (chainOfCommand.status === "approved_once" && anomaly.proposedAction === "block") {
        decision = "block";
        escalatedByAnomaly = true;
      }

      if (policy.mode === "monitor") {
        decision = "allow";
      }

      const latencyMs = Number((performance.now() - started).toFixed(3));

      const result: GuardDecisionResult = {
        decision,
        policyDecision,
        findings,
        anomaly,
        eventId: event.eventId,
        policyId: policy.id,
        latencyMs,
        escalatedByAnomaly,
        unsupportedByPolicy,
        chainOfCommand
      };

      if (config.onAudit) {
        try {
          await config.onAudit({ event, result });
        } catch {
          // Audit sink issues must not affect the guarded decision path.
        }
      }

      return result;
    },

    async authorizeUnsupportedCall(
      context: ToolCallContext,
      input: ChainOfCommandAuthorizationInput
    ) {
      if (!chainEnabled) {
        throw new Error("Chain of command is disabled for this guard instance");
      }

      const reviewerId = input.reviewerId?.trim();
      const reason = input.reason?.trim();
      if (!reviewerId || !reason) {
        throw new Error("Chain of command approval requires reviewerId and reason");
      }

      if (!Array.isArray(input.reviewTrail) || input.reviewTrail.length === 0) {
        throw new Error("Chain of command approval requires a non-empty reviewTrail");
      }

      const event = normalizeToolCall(context);
      const findings = evaluatePolicy(event, policy);
      const defaultAction = fallbackAction(policy.defaults.action);
      const policyDecision = decisionFromFindings(findings, defaultAction);
      const unsupportedByPolicy = isUnsupportedByPolicy({
        fallbackAction: defaultAction,
        findingsCount: findings.length,
        policyDecision
      });

      if (!unsupportedByPolicy) {
        throw new Error("Chain of command approval is only available for unsupported-by-policy calls");
      }

      const fingerprint = unsupportedCallFingerprint(event);
      const permit = createPermitRecord(event, fingerprint, {
        reviewerId,
        reason,
        supervisorSignature: input.supervisorSignature,
        reviewTrail: input.reviewTrail
      });

      await writePermit(store, permit);
      return permit;
    },

    wrapTool<TInput, TOutput>(
      toolName: string,
      toolFn: (input: TInput, guardResult: GuardDecisionResult) => Promise<TOutput> | TOutput,
      opts: WrapToolOptions<TInput> = {}
    ) {
      return async (input: TInput): Promise<TOutput> => {
        const contextPatch = opts.buildContext?.(input) ?? {};
        const context: ToolCallContext = {
          ...contextPatch,
          toolName,
          args: contextPatch.args ?? input
        };

        let result = await engine.beforeToolCall(context);

        if (
          result.decision === "block" &&
          result.chainOfCommand.status === "eligible" &&
          opts.onChainOfCommandReview
        ) {
          const reviewTrail: ChainOfCommandReviewTrailEntry[] = [];
          let supervisorId = initialSupervisorId(context);

          for (let level = 1; level <= maxEscalationLevels; level += 1) {
            const response = await opts.onChainOfCommandReview({
              level,
              maxLevels: maxEscalationLevels,
              supervisorId,
              input,
              result,
              fingerprint: result.chainOfCommand.fingerprint ?? "",
              reviewTrail: reviewTrail.map((entry) => ({ ...entry }))
            });

            const decision = decisionFromReviewResponse(response);
            if (!decision) {
              result = deniedChainResult(result, reviewTrail);
              break;
            }

            const trailEntry: ChainOfCommandReviewTrailEntry = {
              level,
              supervisorId,
              decision,
              reviewerId: response.reviewerId,
              reason: response.reason,
              supervisorSignature: response.supervisorSignature,
              nextSupervisorId: response.nextSupervisorId,
              ts: new Date().toISOString()
            };

            reviewTrail.push(trailEntry);

            if (decision === "escalate") {
              const nextSupervisorId = response.nextSupervisorId?.trim();
              if (!nextSupervisorId) {
                result = deniedChainResult(result, reviewTrail);
                break;
              }

              supervisorId = nextSupervisorId;
              if (level === maxEscalationLevels) {
                result = deniedChainResult(result, reviewTrail);
              }
              continue;
            }

            const evidence = assertTerminalEvidence(response);
            if (!evidence) {
              result = deniedChainResult(result, reviewTrail);
              break;
            }

            if (decision === "no") {
              result = deniedChainResult(result, reviewTrail);
              break;
            }

            await engine.authorizeUnsupportedCall(context, {
              reviewerId: evidence.reviewerId,
              reason: evidence.reason,
              supervisorSignature: evidence.supervisorSignature,
              reviewTrail
            });

            result = await engine.beforeToolCall(context);
            break;
          }
        }

        if (result.decision === "block") {
          throw new GuardBlockedError(result);
        }

        if (result.decision === "require_approval") {
          if (!opts.onRequireApproval) {
            throw new GuardApprovalRequiredError(result);
          }

          const approved = await opts.onRequireApproval(result, input);
          if (!approved) {
            throw new GuardApprovalDeniedError(result);
          }
        }

        return await toolFn(input, result);
      };
    }
  };

  return engine;
}

File: packages/python/src/twire_guard/guard.py
from __future__ import annotations

import datetime as dt
import inspect
import time
from typing import Any, Callable

from .anomaly import InMemoryStore, get_default_anomaly_config, score_anomaly
from .chain_of_command import (
    chain_of_command_enabled,
    chain_of_command_max_levels,
    consume_permit,
    create_permit_record,
    is_unsupported_by_policy,
    unsupported_call_fingerprint,
    write_permit,
)
from .decision import decision_from_findings, merge_decision_with_anomaly
from .errors import GuardApprovalDeniedError, GuardApprovalRequiredError, GuardBlockedError
from .evaluate import evaluate_policy
from .normalize import normalize_tool_call
from .utils import read_key


def _fallback_action(policy_default: Any) -> str:
    if policy_default in ("allow", "require_approval", "block"):
        return str(policy_default)
    return "allow"


def _decision_from_review_response(response: dict[str, Any]) -> str | None:
    decision = read_key(response, "decision", "decision")
    if decision in ("yes", "no", "escalate"):
        return str(decision)
    return None


def _initial_supervisor_id(context: dict[str, Any]) -> str:
    metadata = context.get("metadata")
    if not isinstance(metadata, dict):
        return "supervisor-level-1"

    value = read_key(metadata, "chain_of_command_supervisor_id", "chainOfCommandSupervisorId")
    if not isinstance(value, str) or value.strip() == "":
        return "supervisor-level-1"
    return value.strip()


def _denied_chain_result(result: dict[str, Any], review_trail: list[dict[str, Any]]) -> dict[str, Any]:
    out = dict(result)
    chain = dict(out.get("chain_of_command", {}))
    chain["status"] = "denied"
    chain["review_trail"] = [dict(entry) for entry in review_trail]
    out["chain_of_command"] = chain
    return out


def _assert_terminal_evidence(response: dict[str, Any]) -> dict[str, Any] | None:
    reviewer_id = read_key(response, "reviewer_id", "reviewerId")
    reason = read_key(response, "reason", "reason")

    reviewer_id_str = reviewer_id.strip() if isinstance(reviewer_id, str) else ""
    reason_str = reason.strip() if isinstance(reason, str) else ""

    if reviewer_id_str == "" or reason_str == "":
        return None

    return {
        "reviewer_id": reviewer_id_str,
        "reason": reason_str,
        "supervisor_signature": read_key(response, "supervisor_signature", "supervisorSignature"),
    }


async def _maybe_await(value: Any) -> Any:
    if inspect.isawaitable(value):
        return await value
    return value


def _normalize_chain_config(config: dict[str, Any]) -> dict[str, Any]:
    chain_config = config.get("chain_of_command")
    if isinstance(chain_config, dict):
        return chain_config

    chain_config = config.get("chainOfCommand")
    if isinstance(chain_config, dict):
        return chain_config

    return {}


def _normalize_anomaly_config(config: dict[str, Any]) -> dict[str, Any]:
    raw = config.get("anomaly")
    if not isinstance(raw, dict):
        return {}

    out = dict(raw)
    aliases = {
        "burstWindowMs": "burst_window_ms",
        "burstMediumCount": "burst_medium_count",
        "burstHighCount": "burst_high_count",
        "zScoreMedium": "zscore_medium",
        "zScoreHigh": "zscore_high",
        "requireApprovalScore": "require_approval_score",
        "blockScore": "block_score",
    }

    for camel, snake in aliases.items():
        if camel in out and snake not in out:
            out[snake] = out[camel]

    return out


class GuardEngine:
    def __init__(self, config: dict[str, Any]) -> None:
        if not isinstance(config, dict):
            raise TypeError("Guard config must be a dictionary")

        policy = config.get("policy")
        if not isinstance(policy, dict):
            raise TypeError("Guard config requires a policy dictionary")

        self.policy = policy
        self.store = config.get("store") if config.get("store") is not None else InMemoryStore()
        self.anomaly_config = {
            **get_default_anomaly_config(),
            **_normalize_anomaly_config(config),
        }

        chain_config = _normalize_chain_config(config)
        enabled_value = read_key(chain_config, "enabled", "enabled")
        max_levels_value = read_key(chain_config, "max_escalation_levels", "maxEscalationLevels")

        self.chain_enabled = chain_of_command_enabled(bool(enabled_value) if enabled_value is not None else None)
        self.max_escalation_levels = chain_of_command_max_levels(max_levels_value)

        on_audit = config.get("on_audit") if config.get("on_audit") is not None else config.get("onAudit")
        self.on_audit = on_audit if callable(on_audit) else None

    async def before_tool_call(self, context: dict[str, Any]) -> dict[str, Any]:
        started = time.perf_counter()
        event = normalize_tool_call(context)
        findings = evaluate_policy(event, self.policy)

        defaults = self.policy.get("defaults") if isinstance(self.policy.get("defaults"), dict) else {}
        default_action = _fallback_action(defaults.get("action"))
        policy_decision = decision_from_findings(findings, default_action)

        unsupported_by_policy = is_unsupported_by_policy(
            {
                "fallback_action": default_action,
                "findings_count": len(findings),
                "policy_decision": policy_decision,
            }
        )

        fingerprint = unsupported_call_fingerprint(event) if unsupported_by_policy else None
        effective_policy_decision = policy_decision
        chain_of_command: dict[str, Any] = {"status": "not_applicable", "review_trail": []}

        if self.chain_enabled and unsupported_by_policy and isinstance(fingerprint, str):
            permit = await consume_permit(
                self.store,
                {"actor_id": event["actor_id"], "session_id": event["session_id"]},
                fingerprint,
            )

            if permit:
                effective_policy_decision = "allow"
                chain_of_command = {
                    "status": "approved_once",
                    "fingerprint": fingerprint,
                    "permit_id": permit.get("permit_id"),
                    "review_trail": [dict(entry) for entry in permit.get("review_trail", [])],
                }
            else:
                chain_of_command = {
                    "status": "eligible",
                    "fingerprint": fingerprint,
                    "review_trail": [],
                }

        anomaly = await score_anomaly(event, self.policy, self.store, self.anomaly_config)

        merged = merge_decision_with_anomaly(effective_policy_decision, str(anomaly["proposed_action"]))
        decision = merged["decision"]
        escalated_by_anomaly = bool(merged["escalated_by_anomaly"])

        if chain_of_command.get("status") == "approved_once" and anomaly.get("proposed_action") == "block":
            decision = "block"
            escalated_by_anomaly = True

        policy_mode = self.policy.get("mode")
        if policy_mode == "monitor":
            decision = "allow"

        latency_ms = round((time.perf_counter() - started) * 1000, 3)

        result = {
            "decision": decision,
            "policy_decision": policy_decision,
            "findings": findings,
            "anomaly": anomaly,
            "event_id": event["event_id"],
            "policy_id": self.policy.get("id"),
            "latency_ms": latency_ms,
            "escalated_by_anomaly": escalated_by_anomaly,
            "unsupported_by_policy": unsupported_by_policy,
            "chain_of_command": chain_of_command,
        }

        if self.on_audit:
            try:
                await _maybe_await(self.on_audit({"event": event, "result": result}))
            except Exception:
                pass

        return result

    async def authorize_unsupported_call(self, context: dict[str, Any], input_value: dict[str, Any]) -> dict[str, Any]:
        if not self.chain_enabled:
            raise ValueError("Chain of command is disabled for this guard instance")

        reviewer_id_raw = read_key(input_value, "reviewer_id", "reviewerId")
        reason_raw = read_key(input_value, "reason", "reason")

        reviewer_id = reviewer_id_raw.strip() if isinstance(reviewer_id_raw, str) else ""
        reason = reason_raw.strip() if isinstance(reason_raw, str) else ""

        if reviewer_id == "" or reason == "":
            raise ValueError("Chain of command approval requires reviewer_id and reason")

        review_trail = read_key(input_value, "review_trail", "reviewTrail")
        if not isinstance(review_trail, list) or len(review_trail) == 0:
            raise ValueError("Chain of command approval requires a non-empty review_trail")

        event = normalize_tool_call(context)
        findings = evaluate_policy(event, self.policy)

        defaults = self.policy.get("defaults") if isinstance(self.policy.get("defaults"), dict) else {}
        default_action = _fallback_action(defaults.get("action"))
        policy_decision = decision_from_findings(findings, default_action)

        unsupported_by_policy = is_unsupported_by_policy(
            {
                "fallback_action": default_action,
                "findings_count": len(findings),
                "policy_decision": policy_decision,
            }
        )

        if not unsupported_by_policy:
            raise ValueError("Chain of command approval is only available for unsupported-by-policy calls")

        fingerprint = unsupported_call_fingerprint(event)
        permit = create_permit_record(
            event,
            fingerprint,
            {
                "reviewer_id": reviewer_id,
                "reason": reason,
                "supervisor_signature": read_key(input_value, "supervisor_signature", "supervisorSignature"),
                "review_trail": [dict(entry) for entry in review_trail if isinstance(entry, dict)],
            },
        )

        await write_permit(self.store, permit)
        return permit

    def wrap_tool(
        self,
        tool_name: str,
        tool_fn: Callable[[Any, dict[str, Any]], Any],
        opts: dict[str, Any] | None = None,
    ) -> Callable[[Any], Any]:
        options = opts or {}

        async def wrapped(input_value: Any):
            build_context = options.get("build_context") or options.get("buildContext")
            context_patch = await _maybe_await(build_context(input_value)) if callable(build_context) else {}
            if not isinstance(context_patch, dict):
                context_patch = {}

            context = {
                **context_patch,
                "tool_name": tool_name,
                "args": context_patch["args"] if "args" in context_patch else input_value,
            }

            result = await self.before_tool_call(context)

            review_cb = options.get("on_chain_of_command_review") or options.get("onChainOfCommandReview")

            if (
                result.get("decision") == "block"
                and result.get("chain_of_command", {}).get("status") == "eligible"
                and callable(review_cb)
            ):
                review_trail: list[dict[str, Any]] = []
                supervisor_id = _initial_supervisor_id(context)

                for level in range(1, self.max_escalation_levels + 1):
                    response_value = await _maybe_await(
                        review_cb(
                            {
                                "level": level,
                                "max_levels": self.max_escalation_levels,
                                "supervisor_id": supervisor_id,
                                "input": input_value,
                                "result": result,
                                "fingerprint": result.get("chain_of_command", {}).get("fingerprint", ""),
                                "review_trail": [dict(entry) for entry in review_trail],
                            }
                        )
                    )

                    response = response_value if isinstance(response_value, dict) else {}
                    decision = _decision_from_review_response(response)
                    if decision is None:
                        result = _denied_chain_result(result, review_trail)
                        break

                    entry = {
                        "level": level,
                        "supervisor_id": supervisor_id,
                        "decision": decision,
                        "reviewer_id": read_key(response, "reviewer_id", "reviewerId"),
                        "reason": read_key(response, "reason", "reason"),
                        "supervisor_signature": read_key(response, "supervisor_signature", "supervisorSignature"),
                        "next_supervisor_id": read_key(response, "next_supervisor_id", "nextSupervisorId"),
                        "ts": dt.datetime.now(dt.timezone.utc).isoformat().replace("+00:00", "Z"),
                    }
                    review_trail.append(entry)

                    if decision == "escalate":
                        next_supervisor_id = read_key(response, "next_supervisor_id", "nextSupervisorId")
                        next_id = next_supervisor_id.strip() if isinstance(next_supervisor_id, str) else ""
                        if next_id == "":
                            result = _denied_chain_result(result, review_trail)
                            break

                        supervisor_id = next_id
                        if level == self.max_escalation_levels:
                            result = _denied_chain_result(result, review_trail)
                        continue

                    evidence = _assert_terminal_evidence(response)
                    if evidence is None:
                        result = _denied_chain_result(result, review_trail)
                        break

                    if decision == "no":
                        result = _denied_chain_result(result, review_trail)
                        break

                    await self.authorize_unsupported_call(
                        context,
                        {
                            "reviewer_id": evidence["reviewer_id"],
                            "reason": evidence["reason"],
                            "supervisor_signature": evidence.get("supervisor_signature"),
                            "review_trail": review_trail,
                        },
                    )

                    result = await self.before_tool_call(context)
                    break

            if result.get("decision") == "block":
                raise GuardBlockedError(result)

            if result.get("decision") == "require_approval":
                on_require_approval = options.get("on_require_approval") or options.get("onRequireApproval")
                if not callable(on_require_approval):
                    raise GuardApprovalRequiredError(result)

                approved = await _maybe_await(on_require_approval(result, input_value))
                if not approved:
                    raise GuardApprovalDeniedError(result)

            return await _maybe_await(tool_fn(input_value, result))

        return wrapped


def create_guard(config: dict[str, Any]) -> GuardEngine:
    return GuardEngine(config)
